src/y0/algorithm/identify/id_star.py

```diff
@@ -0,0 +1,233 @@
+        new_query_variables
+    )
+
+
+def make_parallel_worlds_graph(
+    graph: NxMixedGraph[Variable], worlds: Collection[Collection[Variable]]
+) -> NxMixedGraph[Variable]:
+    """Make Parallel worlds graph"""
+    combined_worlds = [make_parallel_world_graph(graph, world) for world in worlds]
+    return combine_parallel_worlds(graph, combined_worlds, worlds)
+
+
+def make_parallel_world_graph(
+    graph: NxMixedGraph[Variable], treatments: Collection[Variable]
+) -> NxMixedGraph[Variable]:
+    """Make one parallel world based on interventions specified"""
+    pw_graph = graph.intervene(treatments)
+    return NxMixedGraph.from_edges(
+        nodes=[node @ treatments for node in graph.nodes()],
+        directed=[(u @ treatments, v @ treatments) for u, v in pw_graph.directed.edges()],
+        undirected=[
+            (u @ treatments, v @ treatments)
+            for u, v in pw_graph.undirected.edges()
+            if (u not in treatments)
+            and (v not in treatments)
+            and (~u not in treatments)
+            and (~v not in treatments)
+        ],
+    )
+
+
+def to_adj(graph: NxMixedGraph):
+    nodes = graph.nodes()
+    directed = {u: [] for u in nodes}
+    undirected = {u: [] for u in nodes}
+    for u, v in graph.directed.edges():
+        directed[u].append(v)
+    for u, v in graph.undirected.edges():
+        undirected[u].append(v)
+    return nodes, directed, undirected
+
+
+def combine_parallel_worlds(
+    graph: NxMixedGraph[Variable],
+    combined_worlds: Collection[NxMixedGraph[Variable]],
+    worlds: Collection[Collection[Variable]],
+) -> NxMixedGraph[Variable]:
+    """Stitch together parallel worlds through the magic of bidirected edges"""
+    # get all the undirected edges in all the parallel worlds
+    undirected = [(u, v) for world in combined_worlds for u, v in world.undirected.edges()]
+    # Stitch together counterfactual variables with observed variables
+    undirected += [
+        (u, u @ treatments)
+        for treatments in worlds
+        for u in graph.nodes()
+        # Don't add an edge if a variable is intervened on
+        if (u not in treatments) and (~u not in treatments)
+    ]
+    undirected += [
+        (u, v @ treatments)
+        for treatments in worlds
+        for u in graph.nodes()
+        for v in graph.undirected.neighbors(u)
+        # Don't add an edge if a variable is intervened on
+        if (v not in treatments) and (~v not in treatments)
+    ]
+    # Stitch together variables from different counterfactual worlds
+    if len(worlds) > 1:
+        undirected += [
+            (u @ treatments_from_world_1, u @ treatments_from_world_2)
+            for treatments_from_world_1, treatments_from_world_2 in combinations(worlds, 2)
+            for u in graph.nodes()
+            # Don't add an edge if a variable is intervened on in either world.
+            if (u not in treatments_from_world_1)
```

Comment:
is this logic correct? the comment says in either world, but it's 4 conditions and'd together

Reply:
this check is done more than once, maybe better to make an auxiliary function for it too

Reply:
The logic is correct. If u and v have a bidirected edge in the observed world, they have a bidirected edge in a counterfactual world **unless** either have been intervened upon.  There are two ways an intervention could happen. Either intervening to have the same value as in the observed world, or a different value.  I do like the idea of refactoring the duplicate code